### ISSUE TITLE
Custom yolov2-tiny (VOC) to onnx for use in Unity with Barracuda

### ISSUE THREAD

**Issue by michalek-marcin** [Unity-Technologies/barracuda-release#152]
Hi,

I found project https://github.com/keijiro/TinyYOLOv2Barracuda which works great with onnx model used in this sample project.

I'm trying to use my custom trained yolov2-tiny model in Unity with Barracuda. I trained my model with darknet and convert to ONNX using this tools:
https://github.com/hollance/YOLO-CoreML-MPSNNGraph
https://github.com/onnx/onnxmltools

One thing is very weird for me. I made two onnx files using exactly same process. First one is original darknet yolov2-tiny (VOC) model. Second one is my custom model trained on the first one. My custom model can detect 3 classes and have same anchors. First model (original yolov2) works in project I mentioned above. Second one doesn't. I made changes in configuration files in project: changed class number, labels (Common.hlsl, Config.cs).

Here you can find my onnx files: https://drive.google.com/drive/folders/1f9To24q4fi9Zyaos4RvrQEYmicvi91xy?usp=sharing
Here are my darknet weights and cfg file: https://drive.google.com/drive/folders/1iSu6S0PCJBYEuokYG7WlPFaoKy729bfy?usp=sharing

Application should detect those buildings:
![emirates3](https://user-images.githubusercontent.com/54703654/109392518-bf45cd00-791c-11eb-977b-712c802661cd.jpg)

@AlexRibard I found this issue and tried to convert from keras to onnx with additional parameters like you suggest in this issue: https://github.com/Unity-Technologies/barracuda-release/issues/99. But result is the same.

UPDATE:
I made test script using onnxruntime. You can find this script inside google drive folder I shared above. Please execute this script with my two onnx files in same folder and image I paste here. My custom model returns output like this:
[[[[nan nan nan ... nan nan nan]
   [nan nan nan ... nan nan nan]
   [nan nan nan ... nan nan nan]
   ...

**Comment by AlexRibard** [comment]
@michalek-marcin I've checked both of your models. They import correctly.
And `yolov2_voc` is executing correctly (validated with onnxruntime)
However for `yolov2_custom` it runs correctly, but as you mention outputs NaN. I've checked with onnxruntime and they also output NaN.
After investigation layer `batchnorm1` `scale` and `B` both have contain NaNs. Also a lot of other batchnorm layer also contains NaN in their scale and bias tensor.

I would check your training code to see if it contains NaN or if the problem comes when exporting.

**Comment by AlexRibard** [comment]
closing since the onnx model contains NaN and onnxruntime exhibit the same inference behaviour

**Comment by michalek-marcin** [comment]
@AlexRibard I finally found reason: https://github.com/allanzelener/YAD2K/issues/65. I had to correct one script which convert my darknet model to keras.

**Comment by AlexRibard** [comment]
ok great to hear it :)

**Comment by getarobo** [comment]
@michalek-marcin I am trying to do exactly what you mentioned and glad it worked out! If you don't mind, would you be kind enough to give some tips or advices or even scripts before go deep into this rabbit hole? Thanks!

**Comment by michalek-marcin** [comment]
Hi @getarobo,

If you have any problems with conversion from darknet to onnx, please follow this steps:

1. Use this tool: [https://github.com/hollance/YOLO-CoreML-MPSNNGraph](url)
  1.1 Use yad2k.py script for convert from darknet to Keras:
`python3 yad2k.py -p path_to_cfg path_to_weights save_path_for_h5_file`
  1.2 Use coreml.py script to convert from keras to CoreML:
`python coreml.py`
2. Use ONNXMLTOOLS [https://github.com/onnx/onnxmltools](url) to convert CoreML model to ONNX:
  2.1 You must create python script to use onnxmltools. You can use this:
```
import onnxmltools
import coremltools

# Load a Core ML model
coreml_model = coremltools.utils.load_spec('yolo-v2-tiny.mlmodel')

# Convert the Core ML model into ONNX
onnx_model = onnxmltools.convert_coreml(coreml_model, 'Example Model')

# Save as protobuf
onnxmltools.utils.save_model(onnx_model, 'yolo-v2-tiny.onnx')
```

And finally you must execute this script:
`python3 convertml.py`

**Comment by getarobo** [comment]
@michalek-marcin  thank you so much, I will try this right now!!!

**Comment by Ademord** [comment]
related question to this issue, I also using yolo on linux my build is giving me an error I have document on this issue https://forum.unity.com/threads/urgent-shader-loading-problem.1194259/. Any clue what this could be?

**Comment by FlorentGuinier** [comment]
As alexandrk stated it seems the related compute shader could not be compiled, likely because target platform does not allow compute shader. You can validate this by looking at https://docs.unity3d.com/ScriptReference/SystemInfo-supportsComputeShaders.html (or at compile kernel aka *.cs in inspector). FYI we currently have a WIP pixel shader worker in barracuda 2.3.1 (support tensor of rank 4 or less + not all operator yet supported) witch might help if compute shader are not supported but GPU is along pixel shader. If the VM do note have access to gpu then the best bet is the Burst backend. However for heavy load such as convolution on images, it will probably be an order of magnitude slower than a GPU (witch typically have at least an order of magnitude more FLOPS).
